src/compute-client/src/controller/mod.rs

```diff
@@ -0,0 +1,474 @@
+    /// compromise the correctness of the system.
+    ///
+    /// This method is **not** guaranteed to be cancellation safe. It **must**
+    /// be awaited to completion.
+    pub async fn process(&mut self) -> Result<Option<ComputeControllerResponse<T>>, anyhow::Error> {
+        match self.compute.stashed_response.take() {
+            None => Ok(None),
+            Some(ActiveReplicationResponse::ComputeResponse(response)) => match response {
+                ComputeResponse::FrontierUppers(updates) => {
+                    let cmd = durable::Command::UpdateWriteFrontiers { updates };
+                    self.apply_durable_command(cmd).await?;
+                    Ok(None)
+                }
+                ComputeResponse::PeekResponse(uuid, peek_response, otel_ctx) => {
+                    let cmd = durable::Command::RemovePeeks {
+                        uuids: [uuid].into(),
+                    };
+                    self.apply_durable_command(cmd).await?;
+                    Ok(Some(ComputeControllerResponse::PeekResponse(
+                        uuid,
+                        peek_response,
+                        otel_ctx,
+                    )))
+                }
+                ComputeResponse::TailResponse(global_id, response) => {
+                    let new_upper = match &response {
+                        TailResponse::Batch(TailBatch { lower, upper, .. }) => {
+                            // Ensure there are no gaps in the tail stream we receive.
+                            assert_eq!(lower, &self.compute.collection(global_id)?.write_frontier);
+
+                            upper.clone()
+                        }
+                        // The tail will not be written to again, but we should not confuse that
+                        // with the source of the TAIL being complete through this time.
+                        TailResponse::DroppedAt(_) => Antichain::new(),
+                    };
+
+                    let updates = vec![(global_id, new_upper)];
+                    let cmd = durable::Command::UpdateWriteFrontiers { updates };
+                    self.apply_durable_command(cmd).await?;
+
+                    Ok(Some(ComputeControllerResponse::TailResponse(
+                        global_id, response,
+                    )))
+                }
+            },
+            Some(ActiveReplicationResponse::ReplicaHeartbeat(replica_id, when)) => Ok(Some(
+                ComputeControllerResponse::ReplicaHeartbeat(replica_id, when),
+            )),
+        }
+    }
+}
+
+// Internal interface
+
+impl<T> ActiveComputeController<'_, T>
+where
+    T: Timestamp + Lattice,
+    ComputeGrpcClient: ComputeClient<T>,
+{
+    async fn apply_durable_command(
+        &mut self,
+        cmd: durable::Command<T>,
+    ) -> Result<(), ComputeError> {
+        self.compute
+            .durable
+            .activate(self.storage_controller)
```

Comment:
I didnt fully process the distinction of DurableState and ActiveDurableState. Can we just create an ActiveDurableState once we have an ActiveComputeController ?

Reply:
`ActiveDurableState` is a wrapper around a `DurableState` bundled with a handle to a storage controller. It's the same relationship as `ComputeController` and `ActiveComputeController`. That pattern exists because we sometimes need mutable access to both the stuff inside `ComputeController` and the storage controller at the same time, so we can't permanently bundle them together. Instead, the compute controller gets "activated" only on demand.

I think it's plausible that the `ActiveComputeController` could reference an `ActiveDurableState` directly. Let me try and see how that works out!